tarea de la clase 7

fngonzalez opened the pull request.

Commit:
Hace todo el archivo comentario

Necesitaba que el archivo se "ignore" para poder trabajar sin errores en la consola

Commit:
Sumar ID para agarrar elemento

Necesito poder agarrar el elemento para poder usarlo

Commit:
Hace parte de la tarea

Commit:
Hago que se usen las validaciones

Commit:
Cambia las funciones a donde corresponden

Moví las funciones a "pruebas" porque así estaba dicho en la clase.

Commit:
cambia las funciones a pruebas

necesito escribir las pruebas para saber si funciona bien el código que escribo

Commit:
Cambia el órden

Se vuelve un poco más visual el código para una mejor lectura

Commit:
Pone las cosas en su lugar y borra cosas

Reordena el código y borré cosas que no iban

Commit:
Da lugar para que se abra el archivo de las validaciones

Commit:
Crea las validaciones para la provincia y el regalo de usuario

Commit:
Crea las pruebas de las validaciones

necesito ver si las funciones que creé para validar, lo están haciendo

Commit:
Cambia de archivo el código

Commit:
Hace que se valide la totalidad de validarProvincia

En la versión anterior solamente se probaba lo que pasaba cuando había algo distinto al nombre de las provincias, siendo que no se probaba cuando se tomaban los valores de las provincias del país

Commit:
Arregla bugs

Commit:
Crea una validación más

Commit:
Arregla un bug

Commit:
Agrega pruebas

Necesito saber que se pruebe el happy case y agrego un caso más para saber que los valores escritos son mayores a dos caracteres.

Commit:
Cambia el string que devuelve el console.assert

fsodano commented on js-xmas-edition-master/index.html:
```diff
@@ -12,101 +13,108 @@
+                    <option value="Catamarca">Catamarca</option>
+                    <option value="Chaco">Chaco</option>
+                    <option value="Chubut">Chubut</option>
+                    <option value="Crdoba">Cordoba</option>
+                    <option value="Corrientes">Corrientes</option>
+                    <option value="Entre Rios">Entre Rios</option>
+                    <option value="Frmosa">Formosa</option>
```
Y eso por  qué?

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,51 @@
+    }
+    else if (regalo.length <= 2) {
+        return "Este campo debe contener al menos más de 2 caracteres"
+    }
```
con el <=2 ya estás cubriendo el caso de === 0

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,51 @@
+        return "Este campo debe contener al menos más de 2 caracteres"
+    }
+    else if (regalo.length > 250) {
+        return "Este campo debe tener menos de 250 caracteres"
```
menor o igual (según lo que dice tu IF)

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,51 @@
+        || provincia == "Sgo. del Ester"
+        || provincia == "Tierra del Fuego"
+        || provincia == "Tucuman"
+    ) {
```
acá te recomiendo que agarres todos los `<option>` y los recorras, y compares si la provincia es igual al valor de alguno de los options, porque te puede pasar que te comas opciones (de movida veo que te comiste córdoba)

Commit:
crea funciones para que el código sea más legible

tuve que hacer funciones nuevas para que las que ya tenía se puedan leer mejor.

Commit:
cambia un texto de validacion

no me gustaba el anterior

Commit:
arregla bug

la funcion no tomaba a la última provincia

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,45 @@
+        arrayFinal.push(arrayInicial[i].value)
+    }
+    return arrayFinal
+}
```
Un par de cosas:

1. en vez de pasar el arrayLength como parámetro, sacalo de arrayInicial.length, que es de todas formas, lo que estás haciendo acá (https://github.com/nahuel-fed/JSXmas/pull/1/files#diff-0da18916537aa9315a920a6cc39c5719R13)
2. Si bien no está mal pasar un arrayFinal como parámetro si tu intención es AGREGAR al mismo array, en realidad en el único lugar que la llamás, le estás pasando un array vacío. O sea que podrías directamente declarar el arrayFinal dentro de esta función.
3. Le ponés "juntar provincias" como nombre de función, pero lo único que hace es pasar un nodeList a un array, porque lo que vos le pasás como parámetro es un listado de options. Entonces un mejor nombre sería `convertirNodeListAArray(nodeList){ //... }`. En ningún momento hacés mención a algo específico de una provincia.

Sin embargo, antes de hacer cualquier cambio, leé todos los comentarios que te dejo.

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,45 @@
+    return arrayFinal
+}
+
+function tomarOpciones() {
```
esta función debería llamarse `obtenerArrayProvincias` por ejemplo.

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,45 @@
+
+$form = document.querySelector("#carta-a-santa")
+let provincia = $form.ciudad.value
+let regalo = $form["descripcion-regalo"].value
```
Si vos preguntás estos valores apenas carga la página, siempre van a valer el valor que tienen esos inputs al momento de cargar la página, cuando el usuario cambie el valor de esos inputs, estas variables no se van a actualizar.

fsodano commented on js-xmas-edition-master/js/main.js:
```diff
@@ -0,0 +1,45 @@
+
+
+function validarProvincia(provincia) {
+    arrayDeProvincias = tomarOpciones()
```
`arrayDeProvincias` acá no está definido localmente, deberías ponerle un `let` o un `cost`, si no javascript entiende que es una variable global.

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+        arrayFinal.push(arrayInicial[i].value)
+    }
+    return arrayFinal
 }
```
Conceptualmente, el hecho de que vuelvas a escribir el código de juntarProvincias y tomarOpciones acá, no tiene sentido. Si hacés esto, te vas a tener que acordar de siempre replicar cualquier cambio a la función original juntarProvincias acá, que está mal.

Vos tenés que usar las funciones juntarProvincias y tomarOpciones del archivo original.

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+    return arrayDeProvincias
+}
+
+function pruebaValidacion(funcion, parametro, returnDeLaFuncion, mensajeDeError) {
+    console.assert(funcion(parametro) === returnDeLaFuncion, mensajeDeError)
+}
```
nice 💯

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+function provinciaValida() {
+    arrayDeProvincias = tomarOpciones()
+    let returnValidarProvincia1 = ""
+    let mensajeDeError1 = "No se esta validando que la opcion sea igual al texto"
```
como estas variables existen sólo dentro de la función `provinciaValida` (que debería en realidad, llamarse `probarProvinciaConSeleccionValida`), no te hace falta ponerle el sufijo `1` a las variables.

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+function provinciaInvalida() {
+    provinciaInvalida = 'Oklahoma'
+    returnValidarProvincia2 = 'La provincia seleccionada no es perteneciente a ningún valor'
+    mensajeDeError2 = "La prov no coincide con la validación"
```
estás declarando todas estas variables como globales. Deberían ser locales. Al hacerlas locales no te hace falta ponerles el sufijo `2`.

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+
+
+
+function probarValidarProv() {
```
las letras son gratis! `probarValidarProvincia`

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+
+
+function probarValidarProv() {
+    provinciaValida()
```
probarProvinciaConSeleccionValida()

fsodano commented on js-xmas-edition-master/js/pruebas.js:
```diff
@@ -1,15 +1,65 @@
+
+function probarValidarProv() {
+    provinciaValida()
+    provinciaInvalida()
```
probarProvinciaConSeleccionInvalida()

Commit:
Arregla bugs y crea nuevas funciones

el codigo no respondia al boton, entonces se adecuo todo para que lo pueda hacer. tambien deja de crear variables globales

Commit:
crea la validacion del nombre

me di cuenta que faltaba validar el nombre

Commit:
Crea la prueba de validarRegalo

Necesitaba la prueba de una nueva funcion

Commit:
Arregla un bug diminuto

borra un } innecesario

Commit:
crea pequenhos cambios, y funciona el css

Crea cambios para que el codigo sea mas comod de leer, y tambien para que el css pueda ser utilizado.

Commit:
arregla bug

se iniciaba la página con error en el selector de provincias

Commit:
Ordena de otra manera las pruebas

El cosigo queda mas elegante usando un array en vez de strings separados

Commit:
agrega validacion de numeros y letras

necesitaba restringir los caracteres que se utilizaban en los inputs

Commit:
crea pruebas para las nuevas versiones de las validaciones

Commit:
cambia de usar arrays a objetos

entendi que era buena idea empezar a dejar de lado a los arrays

Commit:
cambia de lugar algunas cosas

solamente quise cambiar el orden

Commit:
arregla bug

necesitaba que tome las direcciones de los formulario, en vez de los formularios en sí

Commit:
hace el codigo mas sucinto

esta mejor, y termina siendo menos largo

Commit:
se manejan los errores desde manejarErrores

los errores que aparecen en rojo, ahora se manejan desde manejarErrores, en vez de manejarlos individualmente desde cada validacion

Commit:
crea funciones para corroboraciones

necesito varias funciones para ver si ciertos parametros estan siendo cumplidos o no

Commit:
deja el codigo mas prolijo

cambios esteticos mas que otra cosa
